ring/ring.go

```diff
@@ -205,268 +287,93 @@ func (r *Ring) setParameters(N int, Modulus []uint64) error {
-
-		// 1.2 Computes Psi and PsiInv in Montgomery form
-		r.NttPsi[i] = make([]uint64, NthRoot>>1)
-		r.NttPsiInv[i] = make([]uint64, NthRoot>>1)
-
-		// Finds a 2N-th primitive Root
-		g := primitiveRoot(qi)
+			qi := tables[i].Modulus
+			bredParams := tables[i].BRedParams
 
-		power := (qi - 1) / NthRoot
-		powerInv := (qi - 1) - power
-
-		// Computes Psi and PsiInv in Montgomery form
-		PsiMont := MForm(ModExp(g, power, qi), qi, r.BredParams[i])
-		PsiInvMont := MForm(ModExp(g, powerInv, qi), qi, r.BredParams[i])
-
-		r.PsiMont[i] = PsiMont
-		r.PsiInvMont[i] = PsiInvMont
-
-		r.NttPsi[i][0] = MForm(1, qi, r.BredParams[i])
-		r.NttPsiInv[i][0] = MForm(1, qi, r.BredParams[i])
-
-		// Computes nttPsi[j] = nttPsi[j-1]*Psi and nttPsiInv[j] = nttPsiInv[j-1]*PsiInv
-		for j := uint64(1); j < NthRoot>>1; j++ {
-
-			indexReversePrev := utils.BitReverse64(uint64(j-1), logNthRoot)
-			indexReverseNext := utils.BitReverse64(uint64(j), logNthRoot)
-
-			r.NttPsi[i][indexReverseNext] = MRed(r.NttPsi[i][indexReversePrev], PsiMont, qi, r.MredParams[i])
-			r.NttPsiInv[i][indexReverseNext] = MRed(r.NttPsiInv[i][indexReversePrev], PsiInvMont, qi, r.MredParams[i])
+			rescaleParams[j-1][i] = MForm(qi-ModExp(qj, qi-2, qi), qi, bredParams)
 		}
 	}
 
-	r.AllowsNTT = true
-
-	return nil
-}
-
-// primitiveRoot computes the smallest primitive root of the given prime q
-func primitiveRoot(q uint64) (g uint64) {
-
-	notFoundPrimitiveRoot := true
-
-	factors := utils.GetFactors(new(big.Int).SetUint64(q - 1)) //Factor q-1, might be slow
-
-	g = 2
-
-	for notFoundPrimitiveRoot {
-		g++
-		for _, factor := range factors {
-			// if for any factor of q-1, g^(q-1)/factor = 1 mod q, g is not a primitive root
-			if ModExp(g, (q-1)/factor.Uint64(), q) == 1 {
-				notFoundPrimitiveRoot = true
-				break
-			}
-			notFoundPrimitiveRoot = false
-		}
-	}
 	return
 }
 
-// Minimal required information to recover the full ring. Used to import and export the ring.
-type ringParams struct {
-	N       int
-	NthRoot uint64
-	Modulus []uint64
-}
-
-// MarshalBinary encodes the target ring on a slice of bytes.
-func (r *Ring) MarshalBinary() ([]byte, error) {
-
-	parameters := ringParams{r.N, r.NthRoot, r.Modulus}
-
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
-	if err := enc.Encode(parameters); err != nil {
-		return nil, err
-	}
-	return buf.Bytes(), nil
-}
-
-// UnmarshalBinary decodes a slice of bytes on the target Ring.
-func (r *Ring) UnmarshalBinary(data []byte) error {
-
-	parameters := ringParams{}
+// GenNTTParams checks that N has been correctly initialized, and checks that each modulus is a prime congruent to 1 mod 2N (i.e. NTT-friendly).
+// Then, it computes the variables required for the NTT. The purpose of ValidateParameters is to validate that the moduli allow the NTT, and to compute the
+// NTT parameters.
+func (r *Ring) GenNTTParams(NthRoot uint64, primitiveRoots []uint64, factors [][]uint64) (err error) {
```

Comment:
Should not be public. `Ring` structs are created entirely with the `NewRing` method.